server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -421,12 +416,46 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
                 snapshotId,
                 ActionListener.map(listener, v -> {
                     cleanupStaleIndices(foundIndices, survivingIndices);
+                    cleanupStaleRootFiles(rootBlobs, updatedRepositoryData);
                     return null;
                 })
             );
         }
     }
 
+    private void cleanupStaleRootFiles(Set<String> rootBlobNames, RepositoryData repositoryData) {
+        final Set<String> allSnapshotIds =
+            repositoryData.getAllSnapshotIds().stream().map(SnapshotId::getUUID).collect(Collectors.toSet());
+        final List<String> blobsToDelete = rootBlobNames.stream().filter(
+            blob -> {
+                if (FsBlobContainer.isTempBlobName(blob)) {
+                    return true;
+                }
+                if ((blob.startsWith(SNAPSHOT_PREFIX) || blob.startsWith("meta-")) && blob.endsWith(".dat")) {
+                    final String foundUUID = blob.substring(SNAPSHOT_INDEX_PREFIX.length() - 1, blob.length() - ".dat".length());
```

Comment:
this works because `meta-` has same length as `snap-`, which is quite subtle. Can you create two separate cases (one for `meta-` and one for `snap-`? You can then also assert that `ChecksumBlobStoreFormat.blobName(foundUUID).equals(blob)`.

Reply:
Done + added assertion :)